### PR TITLE
PubSub Notification samples

### PR DESCRIPTION
**Pull request by anshuldavid13**

### CODE REVIEW

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
Looking good overall

File: storage/api/Storage.Samples.Tests/ListPubSubNotificationTest.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        _fixture = fixture;
+    }
+
+    [Fact]
+    public void ListPubSubNotification()
+    {
+        CreatePubSubNotificationSample createPubSubNotificationSample = new CreatePubSubNotificationSample();
+        ListPubSubNotificationSample listPubSubNotificationSample = new ListPubSubNotificationSample();
+
+        var bucketName = Guid.NewGuid().ToString();
```
[the remark this answers is not in the document]
Comment (anshuldavid13) [reply storage/api/Storage.Samples.Tests/ListPubSubNotificationTest.cs]: Used bucketNameGeneric instead

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
More or less the same of questions for all the samples in the PR. We can chat about this.

**Comment by amanda-tarafa** [comment]
@anshuldavid13 there are conflicts on this PR now, I'll wait for those to be fixed bedore reviewing.

File: storage/api/Storage.Samples/GetPubSubNotification.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+using Google.Cloud.Storage.V1;
+using Google.Apis.Storage.v1.Data;
+
+public class GetPubSubNotificationSample
+{
+    public Notification GetPubSubNotification(
+     string bucketName = "your-unique-bucket-name",
+     string notificationId = "notification-Id")
+    {
+        var storage = StorageClient.Create();
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples/GetPubSubNotification.cs:27]: Same comment as before, on sample code we use types.

(For later and not urgent, but if I let some of these slip on the other PRs, do please create a new PR to fix them, thanks)

File: storage/api/Storage.Samples/GetPubSubNotification.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+        Console.WriteLine("ID: " + notification.Id);
+        Console.WriteLine("Topic: " + notification.Topic);
+        Console.WriteLine("EventTypes: " + notification.EventTypes);
+        Console.WriteLine("CustomAttributes: " + notification.CustomAttributes);
+        Console.WriteLine("PayloadFormat: " + notification.PayloadFormat);
+        Console.WriteLine("ObjectNamePrefix: " + notification.ObjectNamePrefix);
+        Console.WriteLine("ETag: " + notification.ETag);
+        Console.WriteLine("SelfLink: " + notification.SelfLink);
+        Console.WriteLine("Kind: " + notification.Kind);
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples/GetPubSubNotification.cs:38]: nit: Maybe there's no need to print these?

File: storage/api/Storage.Samples/GetPubSubNotification.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START storage_print_pubsub_bucket_notification]
+
+using System;
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples/GetPubSubNotification.cs:17]: nit: alphabetically as well here

File: storage/api/Storage.Samples.Tests/GetPubSubNotificationTest.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        GetPubSubNotificationSample getPubSubNotificationSample = new GetPubSubNotificationSample();
+        CreatePubSubNotificationSample createPubSubNotificationSample = new CreatePubSubNotificationSample();
+
+        var topicId = "test" + Guid.NewGuid().ToString();
+        var topic = _fixture.createTopic(topicId);
+
+        var notification = createPubSubNotificationSample.CreatePubSubNotification(_fixture.BucketNameGeneric, topic.Name);
+        var recievedNotification = getPubSubNotificationSample.GetPubSubNotification(_fixture.BucketNameGeneric, notification.Id);
+
+        Assert.NotNull(recievedNotification);
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples.Tests/GetPubSubNotificationTest.cs:40]: Maybe check that the notification.Topic is the one you are expecting.

File: storage/api/Storage.Samples/DeletePubSubNotification.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+// [START storage_delete_bucket_notification]
+
+using System;
+using Google.Cloud.Storage.V1;
+
+public class DeletePubSubNotificationSample
+{
+    public void DeletePubSubNotification(
+    string bucketName = "your-unique-bucket-name",
+    string notificationId = "notificationId")
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples/DeletePubSubNotification.cs:24]: Indent wrong, we indent by 4 spaces. Please review here and everywhere. Also, just use Visual Studio formatter and it should all be corrected.

File: storage/api/Storage.Samples.Tests/DeletePubSubNotificationTest.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+        var topicId = "test" + Guid.NewGuid().ToString();
+        var topic = _fixture.createTopic(topicId);
+
+        var notification = createPubSubNotificationSample.CreatePubSubNotification(_fixture.BucketNameGeneric, topic.Name);
+        deletePubSubNotificationSample.DeletePubSubNotification(_fixture.BucketNameGeneric, notification.Id);
+
+        var exception = Assert.Throws<GoogleApiException>(() => getPubSubNotificationSample.GetPubSubNotification(_fixture.BucketNameGeneric, notification.Id));
+        Assert.Equal(System.Net.HttpStatusCode.NotFound, exception.HttpStatusCode);
+
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples.Tests/DeletePubSubNotificationTest.cs:44]:
```suggestion
```

File: storage/api/Storage.Samples/CreatePubSubNotification.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+// [START storage_create_bucket_notifications]
+
+using Google.Apis.Storage.v1.Data;
+using Google.Cloud.Storage.V1;
+using System;
+
+public class CreatePubSubNotificationSample
+{
+    public Notification CreatePubSubNotification(
+    string bucketName = "your-unique-bucket-name",
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples/CreatePubSubNotification.cs:24]: indent

File: storage/api/Storage.Samples/CreatePubSubNotification.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    string topic = "my-topic")
+    {
+        var storage = StorageClient.Create();
+        Notification notification = new Notification();
+        notification.Topic = topic;
+        notification.PayloadFormat = "JSON_API_V1";
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples/CreatePubSubNotification.cs:30]:
```suggestion
        Notification notification = new Notification
        {
            Topic = topic,
            PayloadFormat = "JSON_API_V1"
        };
```

File: storage/api/Storage.Samples.Tests/CreatePubSubNotificationTest.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    public CreatePubSubNotificationTest(StorageFixture fixture)
+    {
+        _fixture = fixture;
+    }
+
+    [Fact]
+    public void CreatePubSubNotification()
+    {
+
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples.Tests/CreatePubSubNotificationTest.cs:31]:
```suggestion
```

File: storage/api/Storage.Samples.Tests/CreatePubSubNotificationTest.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        CreatePubSubNotificationSample createPubSubNotificationSample = new CreatePubSubNotificationSample();
+        var topicId = "test"+ Guid.NewGuid().ToString().Substring(0, 18);
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples.Tests/CreatePubSubNotificationTest.cs:33]: Why are you grabbing the substring here and not in the rest of tests?

File: storage/api/Storage.Samples.Tests/CreatePubSubNotificationTest.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        var topic = _fixture.createTopic(topicId);
+        var confirmNotification = createPubSubNotificationSample.CreatePubSubNotification(_fixture.BucketNameGeneric, topic.Name);
+
+        Assert.NotNull(confirmNotification);
+
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples.Tests/CreatePubSubNotificationTest.cs:38]:
```suggestion
```

File: storage/api/Storage.Samples.Tests/CreatePubSubNotificationTest.cs [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    [Fact]
+    public void CreatePubSubNotification()
+    {
+
+        CreatePubSubNotificationSample createPubSubNotificationSample = new CreatePubSubNotificationSample();
+        var topicId = "test"+ Guid.NewGuid().ToString().Substring(0, 18);
+        var topic = _fixture.createTopic(topicId);
+        var confirmNotification = createPubSubNotificationSample.CreatePubSubNotification(_fixture.BucketNameGeneric, topic.Name);
+
+        Assert.NotNull(confirmNotification);
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples.Tests/CreatePubSubNotificationTest.cs:37]: Maybe check that the notification topis is the correct one.

File: storage/api/Storage.Samples.Tests/StorageFixture.cs [hunk 14]
```diff
@@ -222,4 +236,19 @@ public void CollectArchivedFiles(string bucketName, string objectName, long? ver
...
+
+    public Topic createTopic(string topicId)
+    {
+        PublisherServiceApiClient publisherClient = PublisherServiceApiClient.Create();
+        TopicName topicName = new TopicName(ProjectId, topicId);
+        var topic = publisherClient.CreateTopic(topicName);
+
+        var policy = new Google.Cloud.Iam.V1.Policy();
+        policy.AddRoleMember("roles/pubsub.publisher", "allUsers");
+        publisherClient.SetIamPolicy(topicName, policy);
```
Comment (amanda-tarafa) [inline storage/api/Storage.Samples.Tests/StorageFixture.cs:248]: Can't you create the topic with the policy set, to try an avoid two backend calls?

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
A few nits and a couple fo questions.

File: storage/api/Storage.Samples.Tests/StorageFixture.cs (hunk 14 above)
Comment (anshuldavid13) [reply storage/api/Storage.Samples.Tests/StorageFixture.cs:249]: I could'nt find another way to do this which incorporates both calls

File: storage/api/Storage.Samples.Tests/StorageFixture.cs [hunk 15]
```diff
@@ -222,4 +236,19 @@ public void CollectArchivedFiles(string bucketName, string objectName, long? ver
...
         versions.Add(version.Value);
     }
+
+    public Topic createTopic(string topicId)
+    {
+        PublisherServiceApiClient publisherClient = PublisherServiceApiClient.Create();
+        TopicName topicName = new TopicName(ProjectId, topicId);
+        var topic = publisherClient.CreateTopic(topicName);
+
+        var policy = new Google.Cloud.Iam.V1.Policy();
```
[the remark this answers is not in the document]
Comment (anshuldavid13) [reply storage/api/Storage.Samples.Tests/StorageFixture.cs:247]: There is a conflict if I am using "using" and hence declared it here

**Review by amanda-tarafa — APPROVED** [review]
LGTM, thanks!

**Pull request merged**